zrml/asset-router/src/pallet_impl/multicurrency.rs

```diff
@@ -46,14 +46,21 @@ impl<T: Config> MultiCurrency<T::AccountId> for Pallet<T> {
     }
 
     fn free_balance(currency_id: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
-        if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-            <T::Currencies as MultiCurrency<T::AccountId>>::free_balance(currency, who)
-        } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
-            T::MarketAssets::reducible_balance(asset, who, false)
+        if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+            // Route "pre new asset system" market assets to `CurrencyType`
+            if T::MarketAssets::asset_exists(asset) {
+                T::MarketAssets::reducible_balance(asset, who, false)
+            } else if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                T::Currencies::free_balance(currency, who)
+            } else {
+                T::MarketAssets::reducible_balance(asset, who, false)
```

Comment:
AFAIU, this is unreachable as every market asset will successfully convert to a currency.

Reply:
If that's correct then I would opt for marking this as unreachable and return a dummy value.

Reply:
This is true under the assumption that all asset types within `MarketAsset` are contained within `Currencies` (as you stated). Should we decide to add another asset type to `MarketAsset` in the future while the lazy migration still is not completed, the code would jump into the dummy value instead of executing the function in `MarketAssets`. For this reason I tend to leave it as is.

Reply:
I see. Yeah, makes sense.